File: src/main/java/algo/backtrack/Subsets_II_90.java
package algo.backtrack;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Subsets_II_90 {
    public static List<List<Integer>> subsetsWithDup(int[] nums) {
        var subsets = new ArrayList<List<Integer>>();
        Arrays.sort(nums);
        backtrack(nums, 0, new ArrayList<>(), subsets);
        return subsets;
    }

    private static void backtrack(int[] nums, int start, List<Integer> set, List<List<Integer>> subsets) {
        if (!subsets.contains(set)) {
            subsets.add(new ArrayList<>(set));
        }
        for (var i = start; i < nums.length; i++) {
            set.add(nums[i]);
            backtrack(nums, i + 1, set, subsets);
            set.remove(set.size() - 1);
        }
    }

    public static void main(String[] args) {
//        var subsets = subsetsWithDup(new int[]{1,2,3});
        var subsets = subsetsWithDup(new int[]{1,2,2});
        subsets.forEach(System.out::println);
    }
}

File: src/main/kotlin/algo/backtrack/Subsets_II_90.kt
package algo.backtrack

class KtSubsets_II_90 {
    fun subsetsWithDup(nums: IntArray): List<List<Int>> {
        val all: MutableList<List<Int>> = mutableListOf()
        nums.sort()
        fun backtrack(nums: IntArray, start: Int, list: MutableList<Int>) {
            all.add(list.toList())
            for (i in start..<nums.size) {
                if (i > start && nums[i] == nums[i - 1]) {
                    continue
                }
                list.add(nums[i])
                backtrack(nums, i + 1, list)
                list.removeLast()
            }
        }
        backtrack(nums, 0, mutableListOf())
        return all
    }
}

fun main() {
    val subset = KtSubsets_II_90()
    println(subset.subsetsWithDup(intArrayOf(1,2,2)))
    println(subset.subsetsWithDup(intArrayOf(0)))
}